druid/src/text/editor.rs

```diff
@@ -0,0 +1,311 @@
+    pub fn drag(&mut self, mouse: &MouseEvent, data: &mut T) {
+        self.do_edit(EditAction::Drag(self.mouse_action_for_event(mouse)), data);
+    }
+
+    /// Handle a copy command
+    pub fn copy(&self, data: &mut T) {
+        if !self.data_is_stale(data) {
+            self.set_clipboard()
+        }
+    }
+
+    /// Handle a cut command
+    pub fn cut(&mut self, data: &mut T) {
+        if !self.data_is_stale(data) {
+            self.set_clipboard();
+            self.delete_backward(data);
+        }
+    }
+
+    /// Handle a paste command
+    pub fn paste(&mut self, t: String, data: &mut T) {
+        self.do_edit(EditAction::Paste(t), data)
+    }
+
+    fn mouse_action_for_event(&self, event: &MouseEvent) -> MouseAction {
+        let pos = self.layout.text_position_for_point(event.pos);
+        MouseAction {
+            row: 0,
+            column: pos,
+            mods: event.mods,
+        }
+    }
+
+    /// Update the editor if the data or env has changed.
+    ///
+    /// The widget owning this `Editor` must call this method during its own
+    /// [`update`] call.
+    ///
+    /// [`update`]: ../trait.Widget.html#tymethod.update
+    pub fn update(&mut self, ctx: &mut UpdateCtx, new_data: &T, env: &Env) {
+        if self.data_is_stale(new_data) {
+            self.layout.set_text(new_data.clone());
+            self.selection.constrain_to(new_data);
+            ctx.request_paint();
+        } else if self.layout.needs_rebuild_after_update(ctx) {
+            ctx.request_paint();
+        }
+        self.rebuild_if_needed(ctx.text(), env);
+    }
+
+    /// Must be called in WidgetAdded
+    pub fn rebuild_if_needed(&mut self, factory: &mut PietText, env: &Env) {
+        self.layout.rebuild_if_needed(factory, env);
+    }
+
+    /// Perform an [`EditAction`](enum.EditAction.html).
+    pub fn do_edit(&mut self, edit: EditAction, data: &mut T) {
+        if self.data_is_stale(data) {
+            log::warn!("editor data changed externally, skipping event {:?}", &edit);
+            return;
+        }
+        match edit {
+            EditAction::Insert(chars) | EditAction::Paste(chars) => self.insert(data, &chars),
+            EditAction::Backspace => self.delete_backward(data),
+            EditAction::Delete => self.delete_forward(data),
+            EditAction::JumpDelete(mvmt) | EditAction::JumpBackspace(mvmt) => {
+                let to_delete = if self.selection.is_caret() {
+                    movement(mvmt, self.selection, data, true)
+                } else {
+                    self.selection
+                };
+                data.edit(to_delete.range(), "");
+                self.selection = Selection::caret(to_delete.min());
+            }
+            EditAction::Move(mvmt) => self.selection = movement(mvmt, self.selection, data, false),
+            EditAction::ModifySelection(mvmt) => {
+                self.selection = movement(mvmt, self.selection, data, true)
+            }
```

Comment:
I think you forgot `EditAction::SelectAll`?

Reply:
good catch, thanks!